doc/sphinx/proof-engine/ltac.rst

```diff
@@ -1160,6 +1173,388 @@ Printing |Ltac| tactics
+
+.. coqtop:: in
+
+   Section Iso_axioms.
+
+.. coqtop:: in
+
+   Variables A B C : Set.
+
+.. coqtop:: in
+
+   Axiom Com : A * B = B * A.
+
+   Axiom Ass : A * (B * C) = A * B * C.
+
+   Axiom Cur : (A * B -> C) = (A -> B -> C).
+
+   Axiom Dis : (A -> B * C) = (A -> B) * (A -> C).
+
+   Axiom P_unit : A * unit = A.
+
+   Axiom AR_unit : (A -> unit) = unit.
+
+   Axiom AL_unit : (unit -> A) = A.
+
+.. coqtop:: in
+
+   Lemma Cons : B = C -> A * B = A * C.
+
+   Proof.
+
+   intro Heq; rewrite Heq; reflexivity.
+
+   Qed.
+
+.. coqtop:: in
+
+   End Iso_axioms.
+
+.. coqtop:: in
+
+   Ltac simplify_type ty :=
+   match ty with
+       | ?A * ?B * ?C =>
+           rewrite <- (Ass A B C); try simplify_type_eq
+       | ?A * ?B -> ?C =>
+           rewrite (Cur A B C); try simplify_type_eq
+       | ?A -> ?B * ?C =>
+           rewrite (Dis A B C); try simplify_type_eq
+       | ?A * unit =>
+           rewrite (P_unit A); try simplify_type_eq
+       | unit * ?B =>
+           rewrite (Com unit B); try simplify_type_eq
+       | ?A -> unit =>
+           rewrite (AR_unit A); try simplify_type_eq
+       | unit -> ?B =>
+           rewrite (AL_unit B); try simplify_type_eq
+       | ?A * ?B =>
+           (simplify_type A; try simplify_type_eq) ||
+           (simplify_type B; try simplify_type_eq)
+       | ?A -> ?B =>
+           (simplify_type A; try simplify_type_eq) ||
+           (simplify_type B; try simplify_type_eq)
+   end
+   with simplify_type_eq :=
+   match goal with
+       | |- ?A = ?B => try simplify_type A; try simplify_type B
+   end.
+
+.. coqtop:: in
+
+   Ltac len trm :=
+   match trm with
+       | _ * ?B => let succ := len B in constr:(S succ)
+       | _ => constr:(1)
+   end.
+
+.. coqtop:: in
+
+   Ltac assoc := repeat rewrite <- Ass.
+
+.. coqtop:: in
+
+   Ltac solve_type_eq n :=
+   match goal with
+       | |- ?A = ?A => reflexivity
+       | |- ?A * ?B = ?A * ?C =>
+           apply Cons; let newn := len B in solve_type_eq newn
+       | |- ?A * ?B = ?C =>
+           match eval compute in n with
+               | 1 => fail
+               | _ =>
+                   pattern (A * B) at 1; rewrite Com; assoc; solve_type_eq (pred n)
+           end
+   end.
+
+.. coqtop:: in
+
+   Ltac compare_structure :=
+   match goal with
+       | |- ?A = ?B =>
+           let l1 := len A
+           with l2 := len B in
+               match eval compute in (l1 = l2) with
+                   | ?n = ?n => solve_type_eq n
+               end
+   end.
+
+.. coqtop:: in
+
+   Ltac solve_iso := simplify_type_eq; compare_structure.
+
+The tactic to judge equalities modulo this axiomatization is shown above.
+The algorithm is quite simple. First types are simplified using axioms that
+can be oriented (this is done by ``simplify_type`` and ``simplify_type_eq``).
+The normal forms are sequences of Cartesian products without Cartesian product
+in the left component. These normal forms are then compared modulo permutation
```

Comment:
People seem to love this term "modulo".  It seems to me it's more of a metaphor than anything precise.  I don't what really means.

Reply:
Given an equivalence relation R over a set X, doing something modulo R means that you're working in the quotient X / R.

Reply:
Thanks.